Transformer interface overhaul

vdumoulin opened the pull request:
Fixes #134.

Things left to do:
* Determine if it is okay to assume that data streams with no iteration scheme will produce examples.
* At the moment, transformers don't work on batches if they produce examples, and vice versa. We might want to bake in some default behaviour.
* Test coverage.

vdumoulin commented:
@rizar @dwf I'd like to conclude this saga before Friday, July 17, as this is the last day I'll be available to work on that until the summer school and I'll be presenting Fuel at the summer school.

Do you think it would be possible for you to review this by then?

dwf commented:
Once the checks pass I can review it pretty much immediately.

dwf commented:
Okay, tests pass, but coverage has dropped noticeably, and there are a few Scrutinizer errors. I guess I'll start reviewing.

dwf commented on fuel/schemes.py:
```diff
@@ -95,13 +104,26 @@ class ConcatenatedScheme(IterationScheme):
 
+    Notes
+    -----
+    All schemes must produce the same type of requests (batches or
```
I think a better way of phrasing it is that "Each scheme must produce either batches or examples, but not a mix of both."

dwf replied:
Nevermind, I didn't actually understand what this said.

vdumoulin replied:
Is it clear enough as it is?

dwf replied:
How about "All schemes being concatenated"

dwf commented on fuel/schemes.py:
```diff
@@ -95,13 +104,26 @@ class ConcatenatedScheme(IterationScheme):
+
     """
     def __init__(self, schemes):
+        if not all(scheme.requests_examples == schemes[0].requests_examples
```
Shorter: `len(set(s.requests_examples for s in schemes)) == 1`

dwf commented on fuel/streams.py:
```diff
@@ -37,11 +37,19 @@ class AbstractDataStream(object):
+        if self.iteration_scheme:
+            self.produces_examples = iteration_scheme.requests_examples
+        else:
+            # TODO: is this a valid assumption?
```
Definitely not a valid assumption. But I think it is fair to punt on this. Here is what I suggest (@rizar might have a different opinion):

```python
@property
def produces_examples(self):
    if self.iteration_scheme: 
        return self.iteration_scheme.requests_examples
    elif not hasattr(self, '_produces_examples'):
        raise ValueError("cannot infer type of stream for {} instance; set the "
                                   "produces_examples attribute to True (for example streams "
                                   "or False (for batch streams).".format(self.__class__.__name__))
    else:
        return self._produces_examples

@produces_examples.setter
def produces_examples(self, value):
    if self.iteration_scheme:
        raise ValueError("cannot set produces_examples on {} instance; "
                                   "determined by iteration scheme {}".format(
                                       self.__class__.__name__,
                                       self.iteration_scheme))
    self._produces_examples = value

vdumoulin replied:
I like that.

vdumoulin replied:
See 2970001 for an overview of what changes it entails.

The biggest change from the users' point of view is that they'll have to get used to call `IterableDataset(...).get_example_stream()` instead of `DataStream(IterableDataset(...))`, because the latter would force them to add an additional `stream.produces_example = True` statement afterwards. I think it's a good trade-off.

dwf replied:
I think we might be able to make that assumption for `DataStream`, actually, if it doesn't have an iteration scheme, but AbstractDataStream should make no assumptions I don't think, since that'll probably be what other kinds of streams inherit from (honestly, the more I think about it the more I think that `iteration_scheme` actually belongs in `DataStream` rather than `AbstractDataStream`, but let's leave that for now) -- @rizar or @bartvm might have comments on that too).

dwf replied:
So I guess in concrete terms that would just mean setting `self.produces_examples = True` if there's no iteration scheme in the `DataStream` constructor.

dwf replied:
Okay, going to look at the rest of this PR. :stuck_out_tongue:

dwf commented on fuel/streams.py:
```diff
@@ -150,6 +158,11 @@ class ServerDataStream(AbstractDataStream):
+    sources : tuple of strings
+        The names of the data sources returned by this data stream.
+    produces_examples : bool
+        Whether this data stream produces examples (as opposed to batches
```
Ideally this should eventually be auto-negotiated. Please remember to file a ticket.

vdumoulin replied:
This is something @lamblin also mentioned to me. I just filed the ticket for that.

dwf commented on fuel/streams.py:
```diff
@@ -163,11 +176,16 @@ class ServerDataStream(AbstractDataStream):
         server's end as well.
+    axis_labels : dict, optional
+        Maps source names to tuples of strings describing axis semantics,
+        one per axis. Defaults to `None`, i.e. no information is available.
```
Ditto.

vdumoulin commented:
@dwf thank you for your time, I did not expect you to review this so quickly!

I'll look into bringing the coverage back up tomorrow.

dwf commented:
Not quite done with the review but figured I'd give you what feedback I
could asap. I'll try and finish tonight.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -8,12 +8,19 @@
+    a `transform_example` method (to act on individual examples), or
+    both methods. If a `transform_any` method is defined, it is used
+    for both example and batch inputs. This is useful if the example
+    and batch version of a transformation are the same.
```
Maybe say something about how if you are going from batches to examples or examples to batches, you should override `get_data` instead.

vdumoulin replied:
Will do.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -344,12 +431,20 @@ class Batch(Transformer):
+            kwargs.setdefault(
+                'axis_labels',
+                dict([(source, ('batch',) + labels if labels else None) for
+                      source, labels in iteritems(data_stream.axis_labels)]))
```
Here you should check for `produces_examples` in kwargs, raise an appropriate error if it's there, and pass `produces_examples=False` to super (since it would be weird if the `Batch` transformer didn't always produce batches...)

dwf replied:
In fact, forget what I said about raising an error: there's no need, since you'll get a complaint about duplicate values for keyword anyway. But you should have `produces_examples=False` here.

vdumoulin replied:
Not sure I agree: the `AbstractDataStream` setter for `produces_examples` will complain because there's an iteration scheme for this stream.

I think we should check that the iteration scheme requests batches, though.

dwf replied:
Okay, right, I forgot that we need to account for the iteration scheme.

dwf replied:
Probably good to write a comment to the effect that it gets its `produces_examples` value from `self.iteration_scheme`, and that we expect it to always be true here.

vdumoulin replied:
Will do.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -65,21 +73,30 @@ def get_epoch_iterator(self, **kwargs):
         return super(Transformer, self).get_epoch_iterator(**kwargs)
 
     def get_data(self, request=None):
```
I like the `transform_example`, `transform_batch` setup. Could you write short docstrings for each of those (I don't think we really need a Parameters section, even).

What I kind of don't like is how `transform_any` is effectively a "hidden" part of the protocol because there's no base class method (though it is mentioned in the docs). Though it's kind of ugly, it might be better to throw catch a specific class of Exception (like `class NeedsStreamType(NotImplementedError): pass`), catch it in this method, and then dispatch one or the other of `transform_batch` or `transform_example`.

We could also write a really simple `StreamAgnosticTransformer` (or even just `AgnosticTransformer`) class that has an abstract `transform_any` and then `transform_batch` and `transform_example` implementations that call it, and direct users to subclass that if their transformation for batches and examples is the same.

dwf replied:
(With the current logic, such an `AgnosticTransformer` would still require examples in, examples out or batches in, batches out, but I think that's fine.)

vdumoulin replied:
I like your idea of having an `AgnosticTransformer`. I'll implement it.

dwf replied:
Just note to self - already fixed.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -97,8 +114,9 @@ class Mapping(Transformer):
     """
-    def __init__(self, data_stream, mapping, add_sources=None):
-        super(Mapping, self).__init__(data_stream)
+    def __init__(self, data_stream, produces_examples, mapping,
```
It seems like a pretty fringe use case to be having a `Mapping` that has a different value of `produces_examples` than its data stream, don't you think? What about getting rid of the constructor argument and just doing

```python
@property
def produces_examples(self):
    return self.data_stream.produces_examples
```

(if you go this route, here it would be good to double check that it's not passed as kwargs, as it would have no effect)

vdumoulin replied:
I think what you're suggesting makes sense.

vdumoulin replied:
Although we could simply call the superclass' constructor by passing `self.data_stream.produces_examples`. An error will be raised if someone tries to pass `produces_examples` as a kwarg.

dwf replied:
Actually, we don't even need to do that (and shouldn't) because the
overrided property's non-existent setter will get called. So that will
catch it if someone tries to pass `produces_examples`.

On Thu, Jul 9, 2015 at 1:33 PM, vdumoulin <notifications@github.com> wrote:

> In fuel/transformers/__init__.py
> <https://github.com/mila-udem/fuel/pull/188#discussion_r34287660>:
>
> > @@ -97,8 +114,9 @@ class Mapping(Transformer):
> >          data under source names `add_sources`.
> >
> >      """
> > -    def __init__(self, data_stream, mapping, add_sources=None):
> > -        super(Mapping, self).__init__(data_stream)
> > +    def __init__(self, data_stream, produces_examples, mapping,
>
> Although we could simply call the superclass' constructor by passing
> self.data_stream.produces_examples. An error will be raised if someone
> tries to pass produces_examples as a kwarg.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/fuel/pull/188/files#r34287660>.
>

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -118,8 +136,15 @@ def get_data(self, request=None):
 @add_metaclass(ABCMeta)
-class SingleMapping(Transformer):
-    """Applies a single mapping to multiple sources.
+class SourcewiseTransformer(Transformer):
```
Similar comment about this class and `AgnosticTransformer` etc. I don't know, I'm interested in hearing what you think of that proposal first.

dwf replied:
I would say that `transform_any_source` kinda breaks symmetry with `transform_source_batch` and `transform_source_example`.

vdumoulin replied:
This also makes sense. I'll implement it.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -441,41 +547,38 @@ def sources(self):
 
 
-class Merge(Transformer):
+class Merge(AbstractDataStream):
```
Why is this now an ADS?

vdumoulin replied:
The `close`, `reset` and `next_epoch` methods defined in `Transformer` assume that the transformer wraps only one data stream, so I would need to override those. Plus, `Merge` already overrides `get_epoch_iterator`, which pretty much removes all the benefits from inheriting from `Transformer`. The only advantage of doing so would be to keep an *is a* relationship with `Transformer`, and even then, it's not 100% clear that merging data streams should be counted as a transformation.

What do you think?

dwf replied:
SGTM to leave as is.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -499,13 +602,34 @@ class Merge(Transformer):
             raise ValueError("wrong number of sources given")
         self.sources = sources
 
+    def close(self):
```
Hmm, why are these suddenly necessary?

vdumoulin replied:
See my reply about ADS. `Merge` has a list of data streams, so the default implementation in `Transformer` would crash.

dwf replied:
K nevermind.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -636,19 +765,21 @@ class FilterSources(Transformer):
                              "data_stream.sources")
+        if data_stream.axis_labels:
+            kwargs.setdefault('axis_labels',
+                              dict([(source, labels) for (source, labels)
```
No need for square brackets here

dwf commented:
I think I've more or less now read the entire PR. Sorry it took me so long, it's a big one.

vdumoulin commented:
@dwf I think I addressed all your comments.

You and I have a different definition of "a long time", because I'm pretty satisfied with the review time.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -8,12 +8,24 @@
 
+    Subclasses must define a `transform_batch` method (to act on batches),
+    a `transform_example` method (to act on individual examples), or
+    both methods. If a `transform_any` method is defined, it is used
```
I think this doc is out of date now

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -130,45 +183,112 @@ class SingleMapping(Transformer):
+    """
     def __init__(self, data_stream, **kwargs):
-        super(Flatten, self).__init__(data_stream, **kwargs)
+        super(Flatten, self).__init__(
```
It would be very nice to automatically infer the new axis labels here, since it's a case where we can.

vdumoulin replied:
I think the farthest we can go here is to set the axis label of the flattened axis to `None` or `''`, but other than that I agree.

dwf replied:
What do we typically call anonymous features in tabular datasets?
"features", I thought?
On Jul 9, 2015 2:43 PM, "vdumoulin" <notifications@github.com> wrote:

> In fuel/transformers/__init__.py
> <https://github.com/mila-udem/fuel/pull/188#discussion_r34295028>:
>
> >      def __init__(self, data_stream, **kwargs):
> > -        super(Flatten, self).__init__(data_stream, **kwargs)
> > +        super(Flatten, self).__init__(
>
> I think the farthest we can go here is to set the axis label of the
> flattened axis to None or '', but other than that I agree.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/fuel/pull/188/files#r34295028>.
>

vdumoulin replied:
I vaguely remember @bartvm advocating for an empty string instead of `'features'`, but I don't remember why.

I'm okay with `'features'`, personally.

vdumoulin replied:
This made me realize one thing: `DataStream`'s `axis_label` attribute defaults to `dataset.axis_labels`, no matter what the iteration scheme is. This means that a data stream initialized with an examplewise iteration scheme will potentially have axis labels containing `'batch'`, even though there's no batch axis in the data.

I don't know exactly how to handle that, though.

vdumoulin replied:
Would it be okay to assume that we can simply remove `'batch'` from the axis labels when a dataset is asked individual examples?

lamblin replied:
The discussion about axis labels started [here](https://github.com/mila-udem/fuel/pull/78#discussion_r28370727).
`'features'` does not really add much over `''`, but I don't feel strongly one way or another.
Regarding `'batch'`, I think we can simply remove whatever axis was named `'batch'` when returning examples. Conversely, a Transformer grouping together examples could automatically add a `'batch'` axis.

dwf replied:
I think so. But I'd maybe be conservative about it. Whatever that means...

On Thu, Jul 9, 2015 at 2:57 PM, vdumoulin <notifications@github.com> wrote:

> In fuel/transformers/__init__.py
> <https://github.com/mila-udem/fuel/pull/188#discussion_r34296447>:
>
> >      def __init__(self, data_stream, **kwargs):
> > -        super(Flatten, self).__init__(data_stream, **kwargs)
> > +        super(Flatten, self).__init__(
>
> Would it be okay to assume that we can simply remove 'batch' from the
> axis labels when a dataset is asked individual examples?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/fuel/pull/188/files#r34296447>.
>

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -130,45 +183,112 @@ class SingleMapping(Transformer):
     """Scales and shifts selected sources by scalar quantities.
 
+    Incoming sources will be treated as numpy arrays (i.e. using
+    `numpy.asarray`).
```
Does this cause any monkey business with dtypes if it's a batch of e.g. float32 arrays in a list?

vdumoulin replied:
I don't know, to be honest. I can try to verify that.

vdumoulin replied:
I made the following test:

``` python
>>> import numpy
>>> batch = [numpy.ones(16, dtype='float32') for i in range(32)]
>>> print(numpy.asarray(batch).dtype)
float32
```

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -180,15 +300,20 @@ class ScaleAndShift(SingleMapping):
         self.scale = scale
         self.shift = shift
-        super(ScaleAndShift, self).__init__(data_stream, **kwargs)
+        kwargs.setdefault('axis_labels', data_stream.axis_labels)
```
Is there a case where we want `axis_labels` to be overrideable?

The only case where I can think of that we would want to override is if scale and shift are arrays, and there could be some broadcasting going on. But right now the assumption is that they are scalars...

dwf replied:
I think in cases of doubt like this one it's fine to leave it setdefault.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -200,31 +325,28 @@ def __init__(self, data_stream, dtype, **kwargs):
             dtype = config.floatX
         self.dtype = dtype
-        super(Cast, self).__init__(data_stream, **kwargs)
+        kwargs.setdefault('axis_labels', data_stream.axis_labels)
```
Same question. Seems like an even clearer case where we can force `axis_labels` in the arg list below. I guess we can   argue "caveat emptor" or something but allowing the constructor to do something nonsensical doesn't seem worthwhile.

dwf replied:
(Also note here that they can still manually set the `axis_labels` attribute if they really know what they're doing, but I think it's good if the constructor forces something that's sane 99% of the time)

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -200,31 +325,28 @@ def __init__(self, data_stream, dtype, **kwargs):
     """Force all floating point numpy arrays to be floatX."""
-    def __init__(self, data_stream):
+    def __init__(self, data_stream, **kwargs):
+        kwargs.setdefault('axis_labels', data_stream.axis_labels)
```
Same question.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -238,8 +360,10 @@ class Filter(Transformer):
-    def __init__(self, data_stream, predicate):
-        super(Filter, self).__init__(data_stream)
+    def __init__(self, data_stream, predicate, **kwargs):
+        kwargs.setdefault('axis_labels', data_stream.axis_labels)
```
This one's less clear to me, I think setdefault is fine.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -271,12 +395,17 @@ class Cache(Transformer):
 
     """
-    def __init__(self, data_stream, iteration_scheme):
+    def __init__(self, data_stream, iteration_scheme, **kwargs):
```
No mention of axis labels here... I think it can be forced in the call to super().

Also, not sure how produces_examples should be handled, as I have no idea how this is used. Does it only emit example streams?

vdumoulin replied:
It's requiring an instance of `BatchSizeScheme` as its iteration scheme, so it means `produces_examples=False` is implicitly forced through that restriction. I'll add a comment making this explicit.

vdumoulin replied:
You're right about axis labels, we can safely assume they won't change, as the effect of `Cache` is to change the batch size of a data stream that produces batches.

dwf commented on fuel/transformers/__init__.py:
```diff
@@ -344,12 +473,26 @@ class Batch(Transformer):
+            raise ValueError('the iteration scheme must request batches, '
+                             'not individual examples.')
+        if data_stream.axis_labels:
+            kwargs.setdefault(
```
Same question about whether this should be overrideable. I guess it's a question of flexibility versus sanity checking. Screw it, let's leave it as is.

dwf commented on fuel/transformers/text.py:
```diff
@@ -24,16 +24,23 @@ class NGrams(Transformer):
 
     """
-    def __init__(self, ngram_order, data_stream, target_source='targets'):
+    def __init__(self, ngram_order, data_stream, target_source='targets',
```
Axis labels here?

vdumoulin replied:
I'm not familiar with this transformer. I'd be more comfortable leaving this as it is and open a ticket.

dwf replied:
k

vdumoulin replied:
Ticket opened as #193 .

vdumoulin commented:
@dwf I went ahead and implemented the logic for default axis labels everywhere I could (which covers all classes you pointed out). Since at the moment there isn't anything enforced with axis labels and they're mostly there as documentation tools, I think we can let users override these default values without too much risk.

vdumoulin commented:
Tests are failing in Blocks because `DataStream` doesn't really allow passing no iteration scheme while `produces_examples` isn't set. It usually happens with `IndexableDataset` instances, and the workaround is to call `dataset.get_example_stream()` instead. I'll make a PR on the Blocks side of things to change that.

dwf commented:
What if we defaulted DataStream (but not AbstractDataStream) to be an
example stream, like I proposed? I don't know if you replied to that.

On Fri, Jul 10, 2015 at 4:09 PM, vdumoulin <notifications@github.com> wrote:

> Tests are failing in Blocks because DataStream doesn't really allow
> passing no iteration scheme while produces_examples isn't set. It usually
> happens with IndexableDataset instances, and the workaround is to call
> dataset.get_example_stream() instead. I'll make a PR on the Blocks side
> of things to change that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/fuel/pull/188#issuecomment-120528922>.
>

vdumoulin commented:
Right, I completely forgot about that. It must have gotten buried somewhere in the comments on outdated commits.

I think it's a good idea.

vdumoulin commented:
@dwf The tests pass, woohoo!

dwf commented on fuel/streams.py:
```diff
@@ -101,8 +125,18 @@ class DataStream(AbstractDataStream):
         super(DataStream, self).__init__(**kwargs)
+        # A DataStream with no iteration scheme is considered an example stream
+        # by default
+        if not self.iteration_scheme:
```
Is there ever a case where it would make sense for a `DataStream` to have `produces_examples` set to `False` as a kwarg?

vdumoulin replied:
I can't think of any case.

dwf replied:
Me neither. Ok.

dwf replied:
(Also noticed there is no **kwargs so yeah)

dwf commented on fuel/datasets/text.py:
```diff
@@ -49,8 +49,7 @@ class TextFile(Dataset):
     ...                      preprocess=lower)
-    >>> from fuel.streams import DataStream
-    >>> for data in DataStream(text_data).get_epoch_iterator():
+    >>> for data in text_data.get_example_stream().get_epoch_iterator():
```
Remove this `get_example_stream` I think. Can revert to the old thing.

dwf commented on tests/test_datasets.py:
```diff
@@ -14,7 +14,7 @@
     def setUp(self):
         self.data = [1, 2, 3]
-        self.stream = DataStream(IterableDataset(self.data))
+        self.stream = IterableDataset(self.data).get_example_stream()
```
Do we need get_example_stream at all anymore?

vdumoulin replied:
Good question. The only place I can think of where we can pass no iteration scheme to `DataStream` is for `IterableDataset` datasets. This doesn't make a very predictable interface.

vdumoulin replied:
I'd be inclined to have a convention along the lines of "if you instantiate `DataStream` without giving it an iteration scheme, it'll take the dataset's `example_iteration_scheme` by default"

vdumoulin replied:
That would remove the need for `get_example_stream`, I think.

dwf replied:
SGTM

dwf commented on fuel/streams.py:
```diff
@@ -37,12 +37,36 @@ class AbstractDataStream(object):
+                                 self.iteration_scheme))
+        self._produces_examples = value
+
     def get_data(self, request=None):
```
So I just realized that this method is empty but non-abstract. Just checked with @bartvm, he can't think of any reason for that to be the case either... A subclass inheriting an empty `get_data` would be absolutely useless. Can you add `@abstractmethod`?

vdumoulin replied:
There's the `Merge` transformer which doesn't use `get_data` at all.

dwf replied:
Hmm. That's interesting.
On Jul 11, 2015 5:58 PM, "vdumoulin" <notifications@github.com> wrote:

> In fuel/streams.py
> <https://github.com/mila-udem/fuel/pull/188#discussion_r34417005>:
>
> > +            raise ValueError("cannot infer type of stream for {} instance; "
> > +                             "set the produces_examples attribute to True "
> > +                             "(for example streams) or False (for batch "
> > +                             "streams).".format(self.__class__.__name__))
> > +        else:
> > +            return self._produces_examples
> > +
> > +    @produces_examples.setter
> > +    def produces_examples(self, value):
> > +        if self.iteration_scheme:
> > +            raise ValueError("cannot set produces_examples on {} instance; "
> > +                             "determined by iteration scheme {}".format(
> > +                                 self.__class__.__name__,
> > +                                 self.iteration_scheme))
> > +        self._produces_examples = value
> > +
> >      def get_data(self, request=None):
>
> There's the Merge transformer which doesn't use get_data at all.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/fuel/pull/188/files#r34417005>.
>

vdumoulin replied:
I've just pushed a commit for that. Short story, we can use `AbstractDataStream`'s default `get_epoch_iterator` and simplify the `Merge` implementation so that it uses `get_data`.

rizar commented on fuel/schemes.py:
```diff
@@ -18,6 +18,12 @@ class IterationScheme(object):
 
+    Attributes
+    ----------
+    requests_examples : bool
```
Hmm... it reads like a verb 'requests' and a noun 'examples' and is slightly confusing. 'produces_examples', 'requests_are_examples'?

vdumoulin replied:
'produces_{example,batch}_requests'?

dwf replied:
Simply `example_requests`? To be honest, I find the current name perfectly clear, in the sense that the iteration scheme is in a sense *requesting* examples or batches from the stream.

rizar replied:
With this explanation from David I am fine :)

rizar commented on fuel/transformers/__init__.py:
```diff
@@ -2,18 +2,27 @@
+
+    In all cases, the transformer is expected to have the same output
+    type (example or batch) as its input type.  If the transformer
+    subclass is going from batches to examples or vice versa, it
```
I would also add that overriding `get_data` has to be done when access to `request` is necessary, like in the `Cache` transformer.

dwf replied:
I think I would also remove "In all cases," and say "Typically," but I feel this has lingered long enough. So I'm going to merge, and then send a separate pull request addressing your comment & mine.

dwf replied:
Oh wait -- yours is already addressed! Okay, merging.

dwf commented:
Thanks for the hard work @vdumoulin. I think this is a big improvement.

rizar commented:
+1, this PR is amazing!

vdumoulin commented:
Thank you for your feedback!
